### PR TITLE
Update `demo.rs` to include related errors

### PR DESCRIPTION
**Pull request by indietyp**
## 🌟 What is the purpose of this PR?

This modifies the `demo.rs` to include related errors. This does not change the code in 

## 👂 Open Questions

> The code in the README and demo.rs are now different. Do we want to copy/paste into the README or create two demos?

`demo.rs`, `demo_related.rs`

## 🔗 Related links

* Previous attempt for related errors: https://github.com/indietyp/hash/blob/9e1f8d7943eeeb32d0370b239bd8c23d08bf318a/packages/libs/error-stack/examples/lang.rs

## 🔍 What does this change?

This changes the example to include related errors.

## 📜 Does this require a change to the docs?

Only if we want to mirror them to the `README.md`

## 📹 Demo

image

### CODE REVIEW

File: packages/libs/error-stack/examples/demo.rs [hunk 1]
```diff
@@ -16,14 +16,33 @@ impl fmt::Display for ParseExperimentError {
...
+
+                acc
+            }),
+            (Ok(_), Err(err)) => Err(err),
+            (Err(acc), Ok(_)) => Err(acc),
+            (Err(mut acc), Err(err)) => Err({
+                acc.extend_one(err);
+
+                acc
+            }),
```
Comment (TimDiekmann) [inline packages/libs/error-stack/examples/demo.rs:41]: I think this is more readable
```suggestion
            (Ok(mut acc), Ok(value)) => {
                acc.push(value);

                Ok(acc)
            },
            (Ok(_), Err(err)) => Err(err),
            (Err(acc), Ok(_)) => Err(acc),
            (Err(mut acc), Err(err)) => {
                acc.extend_one(err);

                Err(acc)
            },
```

File: packages/libs/error-stack/examples/demo.rs [hunk 2]
```diff
@@ -16,14 +16,33 @@ impl fmt::Display for ParseExperimentError {
...
+    let values = description
+        .split(' ')
+        .map(|value| {
+            value
+                .parse::<u64>()
+                .into_report()
+                .attach_printable_lazy(|| format!("{value:?} could not be parsed as experiment"))
+        })
+        .map(|value| value.map(|ok| (ok, 2 * ok)))
+        .fold(Ok(vec![]), |acc, value| match (acc, value) {
```
Comment (TimDiekmann) [inline packages/libs/error-stack/examples/demo.rs:29]: Would you mind renaming `acc` to `accum`? I always confuse `acc` with `account` 🙂

File: packages/libs/error-stack/examples/demo.rs [hunk 3]
```diff
@@ -49,21 +68,42 @@ fn start_experiments(
...
+                    acc
+                }),
+                (Ok(_), Err(err)) => Err(err),
+                (Err(acc), Ok(_)) => Err(acc),
+                (Err(mut acc), Err(err)) => Err({
+                    acc.extend_one(err);
+
+                    acc
+                }),
+            },
```
Comment (TimDiekmann) [inline packages/libs/error-stack/examples/demo.rs:97]: Same as above

**Review by TimDiekmann — CHANGES_REQUESTED** [review]
> The code in the README and demo.rs are now different. Do we want to copy/paste into the README or create two demos?

I think we should only have one demo file but until 0.3 we shouldn't replace the demo in the README as the usage is not really straightforward. We could link to this file from the README, however (or to the *examples/* folder).

File: packages/libs/error-stack/examples/demo.rs (hunk 3 above)
Comment (indietyp) [reply packages/libs/error-stack/examples/demo.rs]: fixed in [d1cf758](https://github.com/hashintel/hash/pull/1101/commits/d1cf758932175d95247ed717ca86bc19d0f628ac)

File: packages/libs/error-stack/examples/demo.rs (hunk 2 above)
Comment (indietyp) [reply packages/libs/error-stack/examples/demo.rs]: fixed in [d1cf758](https://github.com/hashintel/hash/pull/1101/commits/d1cf758932175d95247ed717ca86bc19d0f628ac)

File: packages/libs/error-stack/examples/demo.rs (hunk 1 above)
Comment (indietyp) [reply packages/libs/error-stack/examples/demo.rs]: fixed in [d1cf758](https://github.com/hashintel/hash/pull/1101/commits/d1cf758932175d95247ed717ca86bc19d0f628ac)

**Comment by indietyp** [comment]
> We could link to this file from the README, however (or to the examples/ folder).c

I added a reference to the examples folder in [e7849c2](https://github.com/hashintel/hash/pull/1101/commits/e7849c24ca7d33f0f97c8e6b0c2fbd7805b7dba3)

**Review by TimDiekmann — APPROVED** [review]
Great, thank you!

**Pull request merged**